### PR TITLE
[c++] Geometry Dataframe

### PR DESCRIPTION
**Pull request by XanthosXanthopoulos**
This PR adds the new `SOMAGeometryDataFrame` object. This new object requires spatial axes to be determined and uses them internally to index geometry objects stored as a WKB encoded attribute. To differentiate between binary attributes and WKB binary attributes the Arrow schema requires a `dtype` metadata key with the value `WKB`.

### CODE REVIEW

**Review by jp-dark — COMMENTED** [review]
That parts of this I understand look good to me, but I'm not familiar enough with the arrow adapter code to have a good sense of what impact the change might have there.

File: libtiledbsoma/src/utils/arrow_adapter.cc
Comment (johnkerl) [inline libtiledbsoma/src/utils/arrow_adapter.cc:1204]: please use `fmt::format` with `{}` as elsewhere in this file

File: libtiledbsoma/src/utils/arrow_adapter.cc
Comment (johnkerl) [inline libtiledbsoma/src/utils/arrow_adapter.cc:1250]: i'd prefer you move the body of this dimension function above the body of the corresponding attribute funtion

File: libtiledbsoma/src/utils/arrow_adapter.h
Comment (johnkerl) [inline libtiledbsoma/src/utils/arrow_adapter.h:283]: i'd prefer you move the declaration of this dimension function above the corresponding attribute function

File: libtiledbsoma/src/soma/soma_geometry_dataframe.h
Comment (johnkerl) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.h:68]: i see no reason not to use const references here

File: libtiledbsoma/src/soma/soma_geometry_dataframe.h
Comment (johnkerl) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.h:81]: are these example column names legit for geometry dataframes?

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc
Comment (johnkerl) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:58]: i believe you can avoid this by having this constructor take the `ArrowTable` parameters as const references

File: libtiledbsoma/test/unit_soma_geometry_dataframe.cc
Comment (johnkerl) [inline libtiledbsoma/test/unit_soma_geometry_dataframe.cc:87]: copy/paste error?

File: libtiledbsoma/test/unit_soma_geometry_dataframe.cc
Comment (johnkerl) [inline libtiledbsoma/test/unit_soma_geometry_dataframe.cc:90]: copy/paste error?

File: libtiledbsoma/test/unit_soma_geometry_dataframe.cc
Comment (johnkerl) [inline libtiledbsoma/test/unit_soma_geometry_dataframe.cc:100]: please use const references to avoid the moves

File: libtiledbsoma/test/unit_soma_geometry_dataframe.cc
Comment (johnkerl) [inline libtiledbsoma/test/unit_soma_geometry_dataframe.cc:149]: seems like we could use some tests which write & read data -- is this planned for a future PR?

**Review by johnkerl — CHANGES_REQUESTED** [review]

File: libtiledbsoma/src/utils/arrow_adapter.cc
Comment (nguyenv) [inline libtiledbsoma/src/utils/arrow_adapter.cc:1018]: For readability, I think this can be pulled out into a separate `_set_spatial_dimension` (or better name) helper function.

File: libtiledbsoma/src/utils/arrow_adapter.cc
Comment (nguyenv) [inline libtiledbsoma/src/utils/arrow_adapter.cc:981]: Maybe we should `const std::string` some of these in `utils/common.h` like "soma_geometry", "tiledb__internal__", "__min", and "__max"

File: libtiledbsoma/test/unit_soma_geometry_dataframe.cc
Comment (XanthosXanthopoulos) [reply libtiledbsoma/test/unit_soma_geometry_dataframe.cc:149]: Yes, there is #3222 which adds tests for read/write.

File: libtiledbsoma/src/soma/soma_geometry_dataframe.h
Comment (XanthosXanthopoulos) [reply libtiledbsoma/src/soma/soma_geometry_dataframe.h:68]: Agrre on const reference here but will require refactoring all other classes which `tiledb_schema_from_arrow_schema`. Should be done to a follow-up PR maybe?

Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_geometry_dataframe.h:68]: it doesn't need to be done everywhere -- it can be done in places you add

it can also be done all at once in a follow-on

**Review by nguyenv — APPROVED** [review]
I think this look good! Maybe wait for one more approval since it's a larger PR though.

File: libtiledbsoma/src/utils/common.h
Comment (johnkerl) [inline libtiledbsoma/src/utils/common.h:46]:
```suggestion
const std::string SOMA_GEOMETRY_DIMENSION_PREFIX = "tiledb_internal__";
```

**Pull request merged**
